### PR TITLE
Use GA4 API for Analytics data 

### PR DESCRIPTION
**Pull request by Rosa-Fox**
- Presents number of active users using the GA4 realtime API to /realtime-traffic.
- Presents number of users per hour for the past 24 hours (along with total users over past 24 hours) to /traffic-report.
- The data from the above 2 points ^ will be used for the number of active users count and graph.
- /popular-content has been updated to use the GA4 API (Google::Analytics::Data::V1beta::AnalyticsData::Client) as opposed to the UA API (AnalyticsreportingV4::AnalyticsReportingService). 
 
We are in the process of switching from UA to GA4, but GA4 page views has been implemented for most of GOV.UK. This and an issue with our Realtime UA data being unreliable, means that we are now using the GA4 API for most of this dashboard. The /search results still uses the UA Realtime API, however it uses oAuth credentials so they don't clash with the Google service credentials. The service credentials are now shared between content.rb, realtime_traffic.rb and traffic_report.rb. Updating /search to GA4 will be investigated as part of a separate task.

popular-content-ga4

popular-content-error

hourly-traffic-report

realtime-traffic

Paired with @GDSNewt

### CODE REVIEW

File: traffic_report.rb [hunk 1]
```diff
@@ -12,7 +12,7 @@ def visits_per_hour_past_day
     formatted = []
     rows.each do |row|
       formatted << {
-        hour: row[:dimension_values][0][:value],
+        hour: format_hour_value(row[:dimension_values][0][:value]),
         visits: row[:metric_values][0][:value]
```
Comment (gclssvglx) [inline traffic_report.rb]: Can we use `dig` here?

File: content.rb [hunk 2]
```diff
@@ -1,96 +1,71 @@
...
 
   def most_popular_govuk_pages
-    format_response.to_json
+    rows = response_hash[:rows]
+    formatted = []
+    rows.each do |row|
+      row_data = {
+        page_views: row[:metric_values][0][:value],
+        page_path: row[:dimension_values][0][:value],
+        page_title: row[:dimension_values][1][:value]
```
Comment (gclssvglx) [inline content.rb]: Maybe we could `dig` these, but I'm a bit concerned about the `[1]` - how sure are we this won't change? Not a problem, but I wonder what is in `[0]`?

File: content.rb [hunk 3]
```diff
@@ -1,96 +1,71 @@
...
+    formatted.to_json  
   end
 
 private
 
-  def google_analytics_request
-    service.batch_get_reports(analytics_reports)
+  def request
+    ::Google::Analytics::Data::V1beta::RunReportRequest.new({
+      property: "properties/330577055",
```
Comment (gclssvglx) [inline content.rb:42]: Env var maybe?

File: content.rb [hunk 4]
```diff
@@ -1,96 +1,71 @@
...
+      date_ranges: [
+        set_date_range
+      ],
+      dimensions: [set_dimension_path, set_dimension_title],
+      metrics: [set_metric],
+      limit: 10
```
Comment (gclssvglx) [inline content.rb:48]: Very nice!

File: content.rb [hunk 5]
```diff
@@ -1,96 +1,71 @@
...
-          row_data = {
-            page_views: row[:metrics][0][:values][0],
-            page_path: row[:dimensions][0],
-            page_title: row[:dimensions][1]
-          }
-          page_data << row_data
-        end
-      page_data
-    end
+  def set_dimension_title
```
Comment (gclssvglx) [inline content.rb:58]: Should we use `set_dimension` like the method in `traffic_report.rb` for consistency 🤔  Almost feels like there a hidden class here.

Comment (gclssvglx) [reply content.rb:58]: Any value in a common super-class...!

File: realtime_traffic.rb [hunk 6]
```diff
@@ -8,10 +8,17 @@ def initialize
...
 
   def active_users
-    formatted_responses = {
-      active_users_30_minutes: response_hash[:rows][0][:metric_values][0][:value]
-    }
-    formatted_responses.to_json
+    begin
+      formatted_responses = {
+        active_users_30_minutes: response_hash[:rows][0][:metric_values][0][:value]
+      }
```
Comment (gclssvglx) [inline realtime_traffic.rb:14]: See previous suggestion

File: content.rb [hunk 7]
```diff
@@ -9,17 +9,30 @@ def initialize
...
+        }
+      ]
+      formatted.to_json
+    else
+      formatted = []
+      rows.each do |row|
+        row_data = {
+          page_views: row[:metric_values][0][:value],
+          page_path: row[:dimension_values][0][:value],
+          page_title: row[:dimension_values][1][:value]
```
Comment (gclssvglx) [inline content.rb:30]: See previous suggestion

File: traffic_report.rb [hunk 8]
```diff
@@ -8,15 +8,25 @@ def initialize
...
+        hour: "No data available",
+        visits: "No data available"
       }
+      formatted.to_json
+    else
+      formatted = []
+      rows.each do |row|
+        formatted << {
+          hour: format_hour_value(row[:dimension_values][0][:value]),
+          visits: row[:metric_values][0][:value]
```
Comment (gclssvglx) [inline traffic_report.rb:25]: See previous suggestion

File: realtime_traffic.rb [hunk 9]
```diff
@@ -8,10 +8,17 @@ def initialize
...
+    rescue StandardError => e
+      puts "#{e.message}"
+      formatted_responses = {active_users_30_minutes: "No data available"}
+      formatted_responses.to_json
```
Comment (gclssvglx) [inline realtime_traffic.rb:18]: Could this be simpler maybe : `{ active_users_30_minutes: "No data available" }.to_json`?

Yeah, I see what you mean in the commit comment - it is odd you have to do this. Great spot though

**Review by gclssvglx — COMMENTED** [review]
Fantastic work on this @Rosa-Fox  and @GDSNewt - super impressive work.

No big blockers, just a few suggestions on tidying up the code a bit.

File: realtime_traffic.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  attr_reader :client
+  def initialize
+    @client = ::Google::Analytics::Data::V1beta::AnalyticsData::Client.new
+  end
+
+  def active_users
+    formatted_responses = {
+      active_users_30_minutes: response_hash[:rows][0][:metric_values][0][:value]
+    }
+    formatted_responses.to_json
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply realtime_traffic.rb]: 👍

File: realtime_traffic.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+private
+
+  def request
+    { property: "properties/330577055", metrics: [set_metric] }
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply realtime_traffic.rb:27]: The ID hasn't tended to be an ENV var in the GA API projects I have worked on so far, so I think is ok.

File: realtime_traffic.rb (hunk 6 above)
Comment (Rosa-Fox) [reply realtime_traffic.rb:14]: 👍

File: realtime_traffic.rb (hunk 9 above)
Comment (Rosa-Fox) [reply realtime_traffic.rb:18]: 👍

File: traffic_report.rb (hunk 8 above)
Comment (Rosa-Fox) [reply traffic_report.rb:25]: 👍

File: content.rb (hunk 2 above)
Comment (Rosa-Fox) [reply content.rb]: I changed it to:

```
row_data = {
          page_views: row.dig(:metric_values).first.dig(:value),
          page_path: row.dig(:dimension_values).first.dig(:value),
          page_title: row.dig(:dimension_values)[1].dig(:value)
        }
```

yeah the structure it returns is quite nested and horrible!

Comment (gclssvglx) [reply content.rb]: Fantastic - thanks @Rosa-Fox 👍  Not much we can do about the API response, but does seem fragile

File: traffic_report.rb (hunk 1 above)
Comment (Rosa-Fox) [reply traffic_report.rb]: I changed it to `hour: format_hour_value(row.dig(:dimension_values).first.dig(:value))` but its crushed into the very last commit for formatting errors

File: traffic_report.rb [hunk 12]
```diff
@@ -67,4 +67,12 @@ def response
   def response_hash
     response.to_h
   end
+
+  def format_hour_value(hour_data)
+    if (hour_data) == "(other)"
+      "total_visitors_24_hours"
+    else
+      hour_data
+    end
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply traffic_report.rb:85]: 👍

File: traffic_report.rb [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    @client = ::Google::Analytics::Data::V1beta::AnalyticsData::Client.new
+  end
+
+  def visits_per_hour_past_day
+    rows = response_hash[:rows]
+    formatted = []
+    rows.each do |row|
+      formatted << {
+        hour: row[:dimension_values][0][:value],
+        visits: row[:metric_values][0][:value]
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply traffic_report.rb]: I changed to:

```
formatted << {
          hour: format_hour_value(row.dig(:dimension_values).first.dig(:value)),
          visits: row.dig(:metric_values).first.dig(:value)
        }
```

but its crushed into the very last error formatting commit.

File: traffic_report.rb [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    Google::Analytics::Data::V1beta::Dimension.new(
+      name: "hour"
+    )
+  end
+
+  def set_metric
+    Google::Analytics::Data::V1beta::Metric.new(
+      name: "activeUsers"
+    )
+  end
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply traffic_report.rb:55]: I added links to the docs.

File: traffic_report.rb [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+  def set_order
+    Google::Analytics::Data::V1beta::OrderBy.new(
+      {
+        dimension: Google::Analytics::Data::V1beta::OrderBy::DimensionOrderBy.new(
+          {
+            dimension_name: 'hour',
+            order_type: Google::Analytics::Data::V1beta::OrderBy::DimensionOrderBy::OrderType::NUMERIC
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply traffic_report.rb:63]: ah this was a nightmare! It was so hard to decipher which structure the API wanted. Spent ages looking through the docs and source code and eventually [found this post on SO](https://stackoverflow.com/questions/69832484/google-analytics-4-data-api-use-order-by) where `OrderBy` was written in PHP and then translated it to Ruby.

**Comment by Rosa-Fox** [comment]
thanks @gclssvglx. Ready for re-review.

**Comment by gclssvglx** [comment]
Thanks Rosa - will do!

File: content.rb (hunk 5 above)
Comment (gclssvglx) [reply content.rb:58]: Cool - sorry, didn't spot that

File: traffic_report.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (gclssvglx) [reply traffic_report.rb:65]: Sounds like poor Google docs strike again!

**Review by gclssvglx — APPROVED** [review]
Thanks for tackling this - really gnarly stuff, especially the Google API 👏 

Fantastic work @Rosa-Fox and @GDSNewt 🥇

**Pull request merged**